lib/Sender.js

```diff
@@ -41,13 +41,24 @@ class Sender extends EventEmitter {
     dataBuffer.writeUInt16BE(code, 0);
     if (dataBuffer.length > 2) dataBuffer.write(data, 2);
 
-    var self = this;
-    this.messageHandlers.push(function (callback) {
-      self.frameAndSend(0x8, dataBuffer, true, mask);
-      callback();
-      if (typeof cb === 'function') cb();
-    });
-    this.flush();
+    if (this.extensions[PerMessageDeflate.extensionName]) {
+      this.enqueue([this.doClose, [dataBuffer, mask, cb]]);
+    } else {
+      this.doClose(dataBuffer, mask, cb);
+    }
+  }
+
+  /**
+   * Sends a close frame.
+   *
+   * @api private
+   */
+  doClose (data, mask, cb) {
+    this.frameAndSend(0x8, data, true, mask);
```

Comment:
Unrelated but I wonder why the callback is not passed to `this.frameAndSend()` and is ignored in `ping()` and `pong()`.
Anyway I think it's better to keep it as is now preserving the current behavior and maybe change/fix this later in another PR.